### ISSUE TITLE
Unexpected warning passing nullable argument to annotated indexer

### ISSUE THREAD

**Issue by cston** [dotnet/roslyn#47221]
A.cs:
```C#
#nullable enable
public class A
{
    public string? this[string x, string? y] => null;
    public static A F(string? s) => new A();
    public static A F(string x, string? y) => new A();
}
```
B.cs:
```C#
#nullable enable
class B
{
    static void Main()
    {
        string? s = null;
        var a = A.F("");
        _ = a["", s];
    }
}
```

Compile:
```
csc /t:library A.cs
csc /r:A.dll B.cs
```
Result:
```
B.cs(8,19): warning CS8604: Possible null reference argument for parameter 'y' in 'string? A.this[string x, string y]'.
```

**Comment by cston** [comment]
cc @jcouv, @eerhardt

**Comment by jcouv** [comment]
After we get that fixed, let's confirm whether this MetadataAsSource test starts working. If not, we'll want to file an IDE issue.

```
        [Fact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)]
        public async Task AnnotationsInIndexer_Missing()
        {
            var metadata = @"
#nullable enable
public class D
{
    public int this[string? a, string? b] { get { throw null!; } }
}";
            var sourceWithSymbolReference = @"
#nullable enable
class C
{
    void M()
    {
        var obj = new [|D|]();
    }
}";
            var expected = $@"#region {FeaturesResources.Assembly} ReferencedAssembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
// {CodeAnalysisResources.InMemoryAssembly}
#endregion

using System.Reflection;

[DefaultMember(""Item"")]
public class [|D|]
{{
    public D();

    public int this[string a, string b] {{ get; }}
}}";

            using var context = TestContext.Create(
                LanguageNames.CSharp,
                SpecializedCollections.SingletonEnumerable(metadata),
                includeXmlDocComments: false,
                languageVersion: "CSharp8",
                sourceWithSymbolReference: sourceWithSymbolReference,
                metadataLanguageVersion: "CSharp8");

            var navigationSymbol = await context.GetNavigationSymbolAsync();
            var metadataAsSourceFile = await context.GenerateSourceAsync(navigationSymbol);
            TestContext.VerifyResult(metadataAsSourceFile, expected);
        }

        [Fact, Trait(Traits.Feature, Traits.Features.MetadataAsSource)]
        public async Task AnnotationsInIndexer_Present()
        {
            var metadata = @"
#nullable enable
public class D
{
    public string? this[string? a, string? b] { get { throw null!; } }
}";
            var sourceWithSymbolReference = @"
class C
{
    void M()
    {
        var obj = new [|D|]();
    }
}";
            var expected = $@"#region {FeaturesResources.Assembly} ReferencedAssembly, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null
// {CodeAnalysisResources.InMemoryAssembly}
#endregion

#nullable enable

using System.Reflection;

[DefaultMember(""Item"")]
public class [|D|]
{{
    public D();

    public string? this[string? a, string? b] {{ get; }}
}}";

            using var context = TestContext.Create(
                LanguageNames.CSharp,
                SpecializedCollections.SingletonEnumerable(metadata),
                includeXmlDocComments: false,
                languageVersion: "CSharp8",
                sourceWithSymbolReference: sourceWithSymbolReference,
                metadataLanguageVersion: "CSharp8");

            var navigationSymbol = await context.GetNavigationSymbolAsync();
            var metadataAsSourceFile = await context.GenerateSourceAsync(navigationSymbol);
            TestContext.VerifyResult(metadataAsSourceFile, expected);
        }
```

**Comment by eerhardt** [comment]
> If not, we'll want to file an IDE issue.

BTW - when I was try to repro this issue myself, VS was constantly crashing on:

```C#
    public abstract partial class Reader 
    {
        public virtual string? this[string name, string? namespaceURI]
        {
            get
            {
                return 
            }
        }
    }
```

and trying to get intellisense to pop up on the end of the `return` line.

**Comment by cston** [comment]
@eerhardt, the IDE crash might be https://github.com/dotnet/roslyn/issues/46622.
